### PR TITLE
wast parser supports type references for func and call_indirect

### PR DESCRIPTION
**Pull request by ColinEberhardt**
just putting in some of the groundwork for #242 ...

### CODE REVIEW

File: types/AST.js [hunk 1]
```diff
@@ -260,8 +260,10 @@ type Func = {
   // Only in WAST
   name: ?Index,
 
-  params: Array<FuncParam>,
-  result: Array<Valtype>,
+  params?: Array<FuncParam>,
+  result?: Array<Valtype>,
+  typeRef?: Identifier | NumberLiteral,
```
Comment (ColinEberhardt) [inline types/AST.js]: Is there a better type to use for this? If not, perhaps we should define a type for this ...

```
type TypeRef = Identifier | NumberLiteral
```

File: types/AST.js [hunk 2]
```diff
@@ -346,6 +348,8 @@ type CallIndirectInstruction = {
   // WAST
   params?: Array<FuncParam>,
   results?: Array<Valtype>,
+  typeRef?: Identifier | NumberLiteral,
```
Comment (ColinEberhardt) [inline types/AST.js]: I'd really like to enforce the constraint that a node has ether `params` & `result` properties, or `typeRef` - is this possible?

File: types/AST.js (hunk 1 above)
Comment (xtuc) [reply types/AST.js]: Yes, there's `Index` https://github.com/xtuc/webassemblyjs/blob/1e06bbd2b3a6117ceca12efc7ed99ef652e4b741/types/AST.js#L46-L54

File: types/AST.js (hunk 2 above)
Comment (xtuc) [reply types/AST.js]: Yes with a variant type. Maybe something like:

```js
type ParamsAndResults = {
   params?: Array<FuncParam>,
   results?: Array<Valtype>
};

type TypeRef = TypeInstruction | ParamsAndResults

type CallIndirectInstruction = {
  ref: TypeRef
};
```

Comment (ColinEberhardt) [reply types/AST.js]: I've expanded on this and outlined in an issue: https://github.com/xtuc/webassemblyjs/issues/257

**Comment by ColinEberhardt** [comment]
This PR has been updated to fix #257 by adding a new `Signature` type.

File: packages/webassemblyjs/src/interpreter/kernel/exec.js [hunk 3]
```diff
@@ -234,9 +234,14 @@ export function executeStackFrame(
          */
         if (typeof func.name === "object") {
           if (func.name.type === "Identifier") {
+            if (func.signature.type !== "Signature") {
+              newRuntimeError(
```
Comment (xtuc) [inline packages/webassemblyjs/src/interpreter/kernel/exec.js]: Just noticed, you're missing a `throw` before. It returns an error.

**Pull request merged**